src/Network/Wai/Middleware/Auth/OAuth2.hs

```diff
@@ -0,0 +1,88 @@
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE RecordWildCards   #-}
+{-# LANGUAGE TemplateHaskell   #-}
+{-# LANGUAGE TupleSections     #-}
+module Network.Wai.Middleware.Auth.OAuth2
+  ( OAuth2(..)
+  , oAuth2Parser
+  ) where
+
+import           Data.Aeson.TH                        (defaultOptions,
+                                                       deriveJSON,
+                                                       fieldLabelModifier)
+import qualified Data.ByteString.Lazy                 as SL
+import           Data.Monoid                          ((<>))
+import qualified Data.Text                            as T
+import           Data.Text.Encoding                   (encodeUtf8)
+import           Network.HTTP.Types                   (status303, status403,
+                                                       status404, status501)
+import qualified Network.OAuth.OAuth2                 as OA2
+import           Network.Wai                          (queryString, responseLBS)
+import           Network.Wai.Auth.Tools               (toLowerUnderscore)
+import           Network.Wai.Middleware.Auth.Provider
+
+
+data OAuth2 = OAuth2
+  { oa2ClientId            :: T.Text
+  , oa2ClientSecret        :: T.Text
+  , oa2AuthorizeEndpoint   :: T.Text
+  , oa2AccessTokenEndpoint :: T.Text
+  , oa2Scope               :: Maybe [T.Text]
+  , oa2ProviderInfo        :: ProviderInfo
+  --, oa2MaxLoginWait :: Maybe Int
+  }
+
+
+oAuth2Parser :: ProviderParser
+oAuth2Parser = mkProviderParser (undefined :: OAuth2)
```

Comment:
I commented above about getting rid of the need for this parameter entirely. However, if it were to stay, it would be better to use a proxy value to avoid the need for an `undefined`.